src/fdc/DiskImageUtils.cc

```diff
@@ -71,30 +78,58 @@ void checkFAT12Partition(SectorAccessibleDisk& disk, unsigned partition)
 struct SetBootSectorResult {
 	unsigned firstDataSector;
 	uint8_t descriptor;
+	bool fat16;
 };
 static SetBootSectorResult setBootSector(
-	MSXBootSector& boot, size_t nbSectors, bool dos1)
+	MSXBootSector& boot, MSXBootSectorType bootType, size_t nbSectors)
 {
 	// start from the default boot block ..
-	const auto& defaultBootBlock = dos1 ? BootBlocks::dos1BootBlock : BootBlocks::dos2BootBlock;
+	const auto& defaultBootBlock = bootType == MSXBootSectorType::DOS1 ? BootBlocks::dos1BootBlock :
+		bootType == MSXBootSectorType::DOS2 ? BootBlocks::dos2BootBlock : BootBlocks::extendedBootBlock;
 	boot = defaultBootBlock.bootSector;
 
 	// .. and fill-in image-size dependent parameters ..
 	// these are the same for most formats
 	uint8_t nbReservedSectors = 1;
 	uint8_t nbHiddenSectors = 1;
+	uint32_t vol_id = random_32bit() & 0x7F7F7F7F; // why are bits masked?;
 
 	// all these are set below (but initialize here to avoid warning)
 	uint16_t nbSides = 2;
 	uint8_t nbFats = 2;
-	uint8_t nbSectorsPerFat = 3;
+	uint16_t nbSectorsPerFat = 3;
 	uint8_t nbSectorsPerCluster = 2;
 	uint16_t nbDirEntry = 112;
 	uint8_t descriptor = 0xF9;
+	bool fat16 = false;
 
 	// now set correct info according to size of image (in sectors!)
-	// and using the same layout as used by Jon in IDEFDISK v 3.1
-	if (nbSectors > 32732) {
+	// and using the same layout as used by Jon in IDEFDISK v 3.1 for FAT12
+	// or Nextor 2.1.1’s FDISK for FAT16.
+	if (nbSectors > 0x10000 && bootType == MSXBootSectorType::EXTENDED) {
+		nbSides = 0;
+		nbFats = 2;
+		nbDirEntry = 512;
+		descriptor = 0xF0;
+		nbHiddenSectors = 0;
+		fat16 = true;
+
+		// <= 128 MB: 4, <= 256 MB: 8, ..., <= 4 GB: 128
+		nbSectorsPerCluster = narrow<uint8_t>(std::clamp(std::bit_ceil(nbSectors) >> 16, size_t(4), size_t(128)));
+
+		// calculate fat size based on cluster count estimate
+		unsigned fatStart = nbReservedSectors + nbDirEntry / DIR_ENTRIES_PER_SECTOR;
+		unsigned estSectorCount = narrow<unsigned>(nbSectors - fatStart);
+		unsigned estClusterCount = std::min(estSectorCount / nbSectorsPerCluster, MAX_FAT16_CLUSTER_COUNT);
```

Comment:
I understand the calculation, but I don't know what `est` stands for (my first idea was 'estimation' but that doesn't make sense).

Reply:
Indeed it’s estimation (as per the comment above). It does not know yet how big the FAT will be, so it calculates the available sector & cluster count without considering space needed for the FAT. This means there is a chance the FAT becomes a sector larger than absolutely needed. After that it calculates the real cluster count taking into account that FAT size, which will always be lower than the prior count so it will still fit in the FAT.

This is the algorithm that Nextor `_FDISK` implements, albeit rewritten a bit for clarity since we don’t need to optimise for Z80.